File: pll.js
//SOURCE: http://www.cubewhiz.com/pll.php

//Only Corner Stuff

//Aa perm
aAPerm = function(){
  algorithm("x (R' U R') D2 (R U' R') D2 R2 x'");
}

//Ab perm
aBPerm = function(){
  algorithm("x R2 D2 (R U R') D2 (R U' R) x'");
}

//E perm
ePerm = function(){
  algorithm("x' (R U') (R' D) (R U R' D') (R U R' D) (R U') (R' D') x");
}

//Only Edge Stuff

//Ua perm
uAPerm = function(){
  algorithm("(R U' R U) (R U) (R U') (R' U' R2)");
}

//Ub perm
uBPerm = function(){
  algorithm("(R2 U) (R U R' U') (R' U') (R' U R')");
}

//H perm
hPerm = function(){
  algorithm("(M2 U) (M2 U2) (M2 U) M2");
}

//Z perm
zPerm = function(){
  algorithm("(M2 U) (M2 U) (M' U2) (M2 U2) (M' U2)");
}

//Swapping two adjacent corners and two edges

//T perm
tPerm = function(){
  algorithm("R U R' U' R' F R2 U' R' U' R U R' F'");
}

//F perm
fPerm = function(){
  algorithm("R' U' F'");
  tPerm();
  front();
  up();
  right();
}

//Ja perm
jAPerm = function(){
  algorithm("R' U L' U2 R U' R' U2 L R U'");
}

//Jb perm
jBPerm = function(){
  algorithm("R U R' F' R U R' U' R' F R2 U' R' U'");
}

//Ra perm
rAPerm = function(){
  algorithm("R U R' F' R U2 R' U2 R' F R U R U2 R' U'");
}

//Rb perm
rBPerm = function(){
  algorithm("R' U2 R U2 R' F R U R' U' R' F' R2 U'");
}

//Three corners and three edges

//Ga perm
gAPerm = function(){
  algorithm("(R2 Uw) (R' U R' U' R Uw') R2 y' (R' U R) y");
}

//Ga perm
gBPerm = function(){
  algorithm("(R' U' R) y (R2 Uw R' U) (R U' R Uw' R2) y'");
}

//Ga perm
gCPerm = function(){
  algorithm("	(R2 Uw' R U') (R U R' Uw R2) (Fw R' Fw')");
}

//Ga perm
gDPerm = function(){
  algorithm("	(R U R') y' (R2 Uw' R U') (R' U R' Uw R2) y");
}

//Two diagonal corners and two edges

//V perm
vPerm = function(){
  algorithm("(R' U R' Dw') (R' F' R2 U') (R' U R' F) (R F) y'");
}

//Na perm
nAPerm = function(){
  algorithm("(z) D (R' U) (R2 D' R D U') (R' U) (R2 D' R U' R) z'");
}

//Nb perm
nBPerm = function(){
  algorithm("	(z) U' (R D') (R2 U R' D U') (R D') (R2 U R' D R') z'");
}

//Y perm
yPerm = function(){
  algorithm("	(F R U') (R' U' R U) (R' F') (R U R' U') (R' F R F')");
}


//Detection Algorithm
findCase = function(){
  let topPieces = [];
  for(var i = 0; i < 4; i++){
    for(var j = 0; j < 3; j++){
      topPieces.push(grid[i][j]);
    }
  }

  //Colors
  let curr = grid[0][4];
  let right = grid[1][4];
  let opposite = grid[2][4];
  let left = grid[3][4];

  //Pll skip case, already solved
  if(topPieces[0] == curr && topPieces[1] == curr && topPieces[2] == curr && topPieces[3] == right && topPieces[4] == right && topPieces[5] == right && topPieces[6] == opposite && topPieces[7] == opposite && topPieces[8] == opposite && topPieces[9] == left && topPieces[10] == left && topPieces[11] == left)
    return true;

  //Diagonal corners --- Possibilities: V perm, Na perm, Nb perm, Y perm
  if(topPieces[0] == curr && topPieces[2] == opposite && topPieces[3] == left && topPieces[5] == right && topPieces[6] == opposite && topPieces[8] == curr && topPieces[9] == right && topPieces[11] == left){
    //V perm
    if(topPieces[4] == opposite && topPieces[7] == right){
      vPerm();
      return true;
    }
    //Nb perm
    else if(topPieces[1] == opposite && topPieces[7] == curr){
      nAPerm();
      return true;
    }
    //Na perm
    else if(topPieces[4] == left && topPieces[10] == right){
      up();
      nBPerm();
      upPrime();
      return true;
    }
    //Y perm
    else if(topPieces[7] == left && topPieces[10] == opposite){
      yPerm();
      return true;
    }
  }

  //Adjacent corners --- Possibilities: Ja perm, Jb perm, T perm, Rb perm, Ra perm, F perm
  if(topPieces[0] == curr && topPieces[2] == right && topPieces[3] == opposite && topPieces[5] == curr && topPieces[6] == right && topPieces[8] == opposite && topPieces[9] == left && topPieces[11] == left){
    //F perm
    if(topPieces[1] == opposite && topPieces[4] == right && topPieces[7] == curr && topPieces[10] == left){
      fPerm();
      return true;
    }
    //T perm
    else if(topPieces[1] == curr && topPieces[4] == left && topPieces[7] == opposite && topPieces[10] == right){
      tPerm();
      return true;
    }
    //Jb perm
    else if(topPieces[1] == right && topPieces[4] == curr && topPieces[7] == opposite && topPieces[10] == left){
      jBPerm();
      return true;
    }

    //Ra perm
    else if(topPieces[1] == curr && topPieces[4] == right && topPieces[7] == left && topPieces[10] == opposite){
      rAPerm();
      return true;
    }
    //Need to turn for these
    //Rb perm
    else if(topPieces[1] == left && topPieces[4] == right && topPieces[7] == opposite && topPieces[10] == curr){
      upPrime();
      rBPerm();
      up();
      return true;
    }
    //Ja perm
    else if(topPieces[1] == curr && topPieces[4] == opposite && topPieces[7] == right && topPieces[10] == left){
      upPrime();
      jAPerm();
      up();
      return true;
    }
  }

  //Edges only --- Possibilities: H perm, Z perm, Ua perm, Ub perm
  if(topPieces[0] == curr && topPieces[2] == curr && topPieces[3] == right && topPieces[5] == right && topPieces[6] == opposite && topPieces[8] == opposite && topPieces[9] == left && topPieces[11] == left){
    //H perm
    if(topPieces[1] == opposite && topPieces[4] == left && topPieces[7] == curr && topPieces[10] == right){
      hPerm();
      return true;
    }
    //Z perm
    if(topPieces[1] == right && topPieces[4] == curr && topPieces[7] == left && topPieces[10] == opposite){
      zPerm();
      return true;
    }
    //Ua perm
    if(topPieces[1] == right && topPieces[4] == left && topPieces[7] == opposite && topPieces[10] == curr){
      uAPerm();
      return true;
    }
    //Ub perm
    if(topPieces[1] == left && topPieces[4] == curr && topPieces[7] == opposite && topPieces[10] == right){
      uBPerm();
      return true;
    }
  }

  //Corners only --- Possibilities: Aa perm, Ab perm, E perm
  if(topPieces[1] == curr && topPieces[4] == right && topPieces[7] == opposite && topPieces[10] == left){

    //Aa perm
    if(topPieces[0] == curr && topPieces[2] == opposite && topPieces[3] == left && topPieces[5] == curr && topPieces[6] == right && topPieces[8] == right && topPieces[9] == opposite && topPieces[11] == left){
      aAPerm();
      return true;
    }

    //Ab perm
    if(topPieces[0] == curr && topPieces[2] == right && topPieces[3] == opposite && topPieces[5] == opposite && topPieces[6] == left && topPieces[8] == curr && topPieces[9] == right && topPieces[11] == left){
      aBPerm();
      return true;
    }
    //E perm
    if(topPieces[0] == left && topPieces[2] == right && topPieces[3] == opposite && topPieces[5] == curr && topPieces[6] == right && topPieces[8] == left && topPieces[9] == curr && topPieces[11] == opposite){
      ePerm();
      return true;
    }
  }

  //Three corners and three edges -- Possibilities: Ga perm, Gb perm
  if(topPieces[1] == curr && topPieces[2] == curr && topPieces[3] == right){
    //Ga perm
    if(topPieces[0] == left && topPieces[4] == opposite && topPieces[5] == left && topPieces[6] == curr && topPieces[7] == left && topPieces[8] == right && topPieces[9] == opposite && topPieces[10] == right && topPieces[11] == opposite){
      gAPerm();
      return true;
    }
    //Gb perm
    if(topPieces[0] == opposite && topPieces[4] == left && topPieces[5] == opposite && topPieces[6] == left && topPieces[7] == right && topPieces[8] == left && topPieces[9] == curr && topPieces[10] == opposite && topPieces[11] == right){
      upPrime();
      gBPerm();
      up();
      return true;
    }
  }
  //Continuation of three corners and three edges -- Possibilities: Gc perm, Gd perm
  if(topPieces[5] == right && topPieces[6] == opposite && topPieces[7] == opposite){
    //Gc perm
    if(topPieces[0] == right && topPieces[1] == left && topPieces[2] == opposite && topPieces[3] == left && topPieces[4] == curr && topPieces[8] == left && topPieces[9] == curr && topPieces[10] == right && topPieces[11] == curr){
      gCPerm();
      return true;
    }
    //Gd perm
    if(topPieces[0] == left && topPieces[1] == right && topPieces[2] == left && topPieces[3] == curr && topPieces[4] == left && topPieces[8] == curr && topPieces[9] == right && topPieces[10] == curr && topPieces[11] == opposite){
      up();
      gDPerm();
      upPrime();
      return true;
    }
    return false;
  }
}

solvePll = function(){
  outer: for(var i = 0; i < 4; i++){
    up();
    for(var j = 0; j < 4; j++){
      if(findCase()){
        break outer;
      }
      turnCubeRight();
    }
  }
}

getAllFunctions = function(){
  var allfunctions=[];
  for (var i in window) {
    if((typeof window[i]).toString()=="function"){
      allfunctions.push(window[i].name);
    }
  }
}

File: solver/pll.py
# SOURCE: http://www.cubewhiz.com/pll.php

from .base import Cube, check_config_matches, Movement

# Only Corner Stuff

# Aa perm
def aa_perm(cube: Cube):
  cube.run_algorithm("x (R' U R') D2 (R U' R') D2 R2 x'")

# Ab perm
def ab_perm(cube: Cube):
  cube.run_algorithm("x R2 D2 (R U R') D2 (R U' R) x'")

# E perm
def e_perm(cube: Cube):
  cube.run_algorithm("x' (R U') (R' D) (R U R' D') (R U R' D) (R U') (R' D') x")

# Only Edge Stuff

# Ua perm
def ua_perm(cube: Cube):
  cube.run_algorithm("(R U' R U) (R U) (R U') (R' U' R2)")

# Ub perm
def ub_perm(cube: Cube):
  cube.run_algorithm("(R2 U) (R U R' U') (R' U') (R' U R')")

# H perm
def h_perm(cube: Cube):
  cube.run_algorithm("(M2 U) (M2 U2) (M2 U) M2")

# Z perm
def z_perm(cube: Cube):
  cube.run_algorithm("(M2 U) (M2 U) (M' U2) (M2 U2) (M' U2)")

# Swapping two adjacent corners and two edges

# T perm
def t_perm(cube: Cube):
  cube.run_algorithm("R U R' U' R' F R2 U' R' U' R U R' F'")

# F perm
def f_perm(cube: Cube):
  cube.run_algorithm("R' U' F' (R U R' U') (R' F) (R2 U') (R' U' R U) (R' U R)")

# Ja perm
def ja_perm(cube: Cube):
  cube.run_algorithm("R' U L' U2 R U' R' U2 L R U'")

# Jb perm
def jb_perm(cube: Cube):
  cube.run_algorithm("R U R' F' R U R' U' R' F R2 U' R' U'")

# Ra perm
def ra_perm(cube: Cube):
  cube.run_algorithm("R U R' F' R U2 R' U2 R' F R U R U2 R' U'")

# Rb perm
def rb_perm(cube: Cube):
  cube.run_algorithm("R' U2 R U2 R' F R U R' U' R' F' R2 U'")

# Three corners and three edges

# Ga perm
def ga_perm(cube: Cube):
  cube.run_algorithm("(R2 Uw) (R' U R' U' R Uw') R2 y' (R' U R) y")

# Ga perm
def gb_perm(cube: Cube):
  cube.run_algorithm("(R' U' R) y (R2 Uw R' U) (R U' R Uw' R2) y'")

# Ga perm
def gc_perm(cube: Cube):
  cube.run_algorithm("(R2 Uw' R U') (R U R' Uw R2) (Fw R' Fw')")

# Ga perm
def gd_perm(cube: Cube):
  cube.run_algorithm("(R U R') y' (R2 Uw' R U') (R' U R' Uw R2) y")

# Two diagonal corners and two edges

# V perm
def v_perm(cube: Cube):
  cube.run_algorithm("(R' U R' Dw') (R' F' R2 U') (R' U R' F) (R F) y'")

# Na perm
def na_perm(cube: Cube):
  cube.run_algorithm("(z) D (R' U) (R2 D' R D U') (R' U) (R2 D' R U' R) z'")

# Nb perm
def nb_perm(cube: Cube):
  cube.run_algorithm("(z) U' (R D') (R2 U R' D U') (R D') (R2 U R' D R') z'")

# Y perm
def y_perm(cube: Cube):
  cube.run_algorithm("(F R U') (R' U' R U) (R' F') (R U R' U') (R' F R F')")


def check_pll_solved(cube: Cube):
  pll_faces = [cube.front, cube.right, cube.back, cube.left]
  return all(face.is_solved() for face in pll_faces)


# Detection Algorithm --- Runs PLL casework to determine which algorithm needs to be applied
# Returns whether or not the PLL case was found. If it was, the solution algorithm should be automatically applied to the cube.
def detect_and_run_pll_case(cube: Cube) -> bool:
  # Pll skip case, already solved
  if check_pll_solved(cube):
    return True

  # Faces / colors used in PLL detection.
  pll_faces = [cube.front, cube.right, cube.back, cube.left]
  front_color, right_color, back_color, left_color = cube.front.center, cube.right.center, cube.back.center, cube.left.center

  top_pieces = cube.front.top_row + cube.right.top_row + cube.back.top_row + cube.left.top_row
  top_corners = [corner for face in pll_faces for corner in (face.top_left, face.top_right)]
  top_edges = [face.top_middle for face in pll_faces]

  # Corners only --- Possibilities: Aa perm, Ab perm, E perm
  # Check that all of the edges are already in the right spot
  if check_config_matches(top_edges, [front_color, right_color, back_color, left_color]):
    if check_config_matches(top_corners, [front_color, back_color, left_color, front_color, right_color, right_color, back_color, left_color]):
      aa_perm(cube)
      return True
    if check_config_matches(top_corners, [front_color, right_color, back_color, back_color, left_color, front_color, right_color, left_color]):
      ab_perm(cube)
      return True
    if check_config_matches(top_corners, [left_color, right_color, back_color, front_color, right_color, left_color, front_color, back_color]):
      e_perm(cube)
      return True

  # Edges only --- Possibilities: H perm, Z perm, Ua perm, Ub perm
  # Check that all of the corners are already in the right spot
  if check_config_matches(top_corners, [front_color, front_color, right_color, right_color, back_color, back_color, left_color, left_color]):
    if check_config_matches(top_edges, [right_color, left_color, back_color, front_color]):
      ua_perm(cube)
      return True
    if check_config_matches(top_edges, [left_color, front_color, back_color, right_color]):
      ub_perm(cube)
      return True
    if check_config_matches(top_edges, [back_color, left_color, front_color, right_color]):
      h_perm(cube)
      return True
    if check_config_matches(top_edges, [right_color, front_color, left_color, back_color]):
      z_perm(cube)
      return True

  # Swapping Two Adjacent Corners & Two Edges --- Possibilities: Ja perm, Jb perm, T perm, Rb perm, Ra perm, F perm
  # Check that the two right corners need to be swapped.
  if check_config_matches(top_corners, [front_color, right_color, back_color, front_color, right_color, back_color, left_color, left_color]):
    if check_config_matches(top_edges, [front_color, back_color, right_color, left_color]):      
      cube.run_algorithm("U'")
      ja_perm(cube)
      cube.run_algorithm("U")
      return True
    if check_config_matches(top_edges, [right_color, front_color, back_color, left_color]):
      jb_perm(cube)
      return True
    if check_config_matches(top_edges, [front_color, left_color, back_color, right_color]):
      t_perm(cube)
      return True
    if check_config_matches(top_edges, [front_color, right_color, back_color, left_color]):
      ja_perm(cube)
      return True
    if check_config_matches(top_edges, [left_color, right_color, back_color, front_color]):
      cube.run_algorithm("U'")
      rb_perm(cube)
      cube.run_algorithm("U")
      return True
    if check_config_matches(top_edges, [front_color, right_color, left_color, back_color]):
      ra_perm(cube)
      return True
    if check_config_matches(top_edges, [back_color, right_color, front_color, left_color]):
      f_perm(cube)
      return True

  # Cycling Three Corners & Three Edges -- Possibilities: Ga perm, Gb perm, Gc perm, Gd perm
  # These don't really have similarities so just check each one individually.
  if check_config_matches(top_pieces, [left_color, front_color, front_color, right_color, back_color, left_color, front_color, left_color, right_color, back_color, right_color, back_color]):
    ga_perm(cube)
    return True
  if check_config_matches(top_pieces, [right_color, left_color, back_color, left_color, right_color, right_color, back_color, front_color, left_color, front_color, back_color, front_color]):
    gb_perm(cube)
    return True
  if check_config_matches(top_pieces, [right_color, left_color, back_color, left_color, front_color, right_color, back_color, back_color, left_color, front_color, right_color, front_color]):
    gc_perm(cube)
    return True
  if check_config_matches(top_pieces, [left_color, back_color, front_color, right_color, right_color, left_color, front_color, left_color, right_color, back_color, front_color, back_color]):
    gd_perm(cube)
    return True

  # Permutations Of Two Diagonal Corners & Two Edges --- Possibilities: V perm, Na perm, Nb perm, Y perm
  # Check that the bottom right and top left corners need to be swapped, and the bottom left and top right corners are in the right spot.
  if check_config_matches(top_corners, [front_color, back_color, left_color, right_color, back_color, front_color, right_color, left_color]):
    if check_config_matches(top_edges, [front_color, back_color, right_color, left_color]):
      v_perm(cube)
      return True
    elif check_config_matches(top_edges, [back_color, right_color, front_color, left_color]):
      na_perm(cube)
      return True
    elif check_config_matches(top_edges, [front_color, left_color, back_color, right_color]):
      cube.run_algorithm("U")
      nb_perm(cube)
      cube.run_algorithm("U'")
      return True
    elif check_config_matches(top_edges, [front_color, right_color, left_color, back_color]):
      y_perm(cube)
      return True

  # None of the configurations matched, return False
  return False
